packages/ui/lit-grid/src/playwright/dx-grid.spec.ts

```diff
@@ -0,0 +1,217 @@
+//
+// Copyright 2021 DXOS.org
+//
+
+import { expect, test, type Browser, type Page, type Locator } from '@playwright/test';
+
+import { setupPage, storybookUrl } from '@dxos/test-utils/playwright';
+
+import { type DxGridCellsSelect, type DxGridPlanePosition } from '../types';
+import { toPlaneCellIndex } from '../util';
+
+const gridPlaneCellSize = 31;
+const gap = 1;
+const nCols = 9;
+const nRows = 7;
+
+class GridManager {
+  constructor(browser: Browser) {
+    this.browser = browser;
+    this.page = null;
+  }
+
+  page: Page | null;
+  browser: Browser;
+
+  async open() {
+    const { page } = await setupPage(this.browser, {
+      url: storybookUrl('dx-grid--spec'),
+      viewportSize: {
+        width: (gridPlaneCellSize + gap) * (nCols + 1.5),
+        height: (gridPlaneCellSize + gap) * (nRows + 1.5),
+      }, // 336 x 272
+    });
+    this.page = page;
+    return this.page.locator('.dx-grid').waitFor({ state: 'visible' });
+  }
+
+  close() {
+    return this.page?.close();
+  }
+
+  planes() {
+    return this.page!.locator('.dx-grid [data-dx-grid-plane]').all();
+  }
+
+  cellsWithinPlane(plane: string) {
+    return this.page!.locator(`.dx-grid [data-dx-grid-plane="${plane}"]`).getByRole('gridcell').all();
+  }
+
+  cell(col: number, row: number, plane: string) {
+    return this.page!.locator(
+      `.dx-grid [data-dx-grid-plane="${plane}"] [aria-colindex="${col}"][aria-rowindex="${row}"]`,
+    );
+  }
+
+  panByWheel(deltaX: number, deltaY: number) {
+    return this.page!.locator('.dx-grid [data-dx-grid-plane="grid"]').dispatchEvent('wheel', { deltaX, deltaY });
+  }
+
+  async forCellsInRange(
+    start: DxGridPlanePosition,
+    end: DxGridPlanePosition,
+    iterator: (col: number, row: number) => Promise<void>,
+  ) {
+    const nCols = 1 + end.col - start.col;
+    const nRows = 1 + end.row - start.row;
+
+    await Promise.all(
+      [...Array(nCols)].map(async (_, c0) => {
+        return Promise.all(
+          [...Array(nRows)].map(async (_, r0) => {
+            return iterator(start.col + c0, start.row + r0);
+          }),
+        );
+      }),
+    );
+  }
+
+  async expectSelectionResult(start: DxGridPlanePosition, end: DxGridPlanePosition) {
+    return this.forCellsInRange(start, end, (col, row) =>
+      expect(this.cell(col, row, 'grid')).toHaveAttribute('aria-selected', 'true'),
+    );
+  }
+
+  async expectVirtualizationResult(cols: number, rows: number, minColIndex = 0, minRowIndex = 0) {
+    await this.cell(minColIndex, minRowIndex, 'grid').waitFor({ state: 'visible' });
+    // Top planes
+    await expect(await this.cellsWithinPlane('fixedStartStart')).toHaveLength(4);
```

Comment:
ime the playwright tests are less flaky when the await is just done by the expect and not beforehand by what is passed into it. Since playwright's expect accepts locators you can expect the locator to return a certain number of items without awaiting `.all()` e.g. https://github.com/dxos/dxos/blob/416409e44fd5c28700586202548a4fb7f9928a85/packages/apps/composer-app/src/playwright/basic.spec.ts#L37

Reply:
I’m not able to remove the await without the tests failing, and removing `.all` changes the API such that `toHaveLength` is no longer available.